### PR TITLE
Add support for iframes

### PR DESCRIPTION
**Pull request by christophemaximin**
Note:
- "Optional" because it is a subclass of `SprocketsLoader`.
- That code was not integrated in `SprocketsLoader` because it's not directly using sprockets, but it does depend on it to load the iframe content assets; also because it uses javascript to load the iframe contents, which can be potentially slow.
- Must be loaded with `Percy::Capybara.use_loader(Percy::Capybara::Loaders::SprocketsWithIframesLoader)`
- If you think it **should** be integrated in `SprocketsLoader`, say the word.

### CODE REVIEW

**Comment by fotinakis** [comment]
First, thanks for this! And for adding tests and all the things, this is great. Does it work when rendered in Percy? :)

Two quick thoughts (just at very quick glance, I will do a more thorough review this week):

1. The ideal world is that this should be default, people shouldn't have to enable it — so that means part of `SprocketsLoader` (and, if you feel up to it, shared in `NativeLoader` as well—but I'm fine with not having support for it there as that is a less common path).
1. Can you add a test for an iframe that is not on the same domain? Selenium often throws terrible errors for single-origin policy violations.
1. Optionally, it should be safe/fast and potentially not require JavaScript. I just discovered a native Selenium method to switch context to the iframe — perhaps you can test out `within_frame` and see if it does what we want? Just need to find a way to get the source while within the frame context, and if Selenium has a native way to do that then it probably will let us dramatically simplify this code and probably not require JavaScript at all — http://stackoverflow.com/a/11910060/128597

**Comment by christophemaximin** [comment]
0- It does work when rendered in Percy! ([example](https://storage.googleapis.com/percy-images/bdfdf904552295142a0b3f83b0ecd0b0210870356c6ced821d84f9d204232cde))
1- I'll look into it for `SprocketsLoader`
2- I'll look into it.
3- I'll look into it.
:)

**Comment by christophemaximin** [comment]
Updates:
1- Now integrated by default in `SprocketsLoader` and `NativeLoader`
2- Added `same-origin` domain check
3- Now using `within_frame` instead of `evaluate_script` to get iframe body

File: lib/percy/capybara/client.rb [hunk 1]
```diff
@@ -63,13 +63,14 @@ def failed?
       end
 
       def initialize_loader(options = {})
+        options[:sprockets_environment] ||= sprockets_environment
```
Comment (fotinakis) [inline lib/percy/capybara/client.rb]: You can revert this change and not pass these into the other loaders anymore.

File: spec/lib/percy/capybara/loaders/base_loader_spec.rb [hunk 2]
```diff
@@ -14,6 +14,18 @@
       expect(resource.content).to include('Hello World!')
       expect(resource.sha).to eq(Digest::SHA256.hexdigest(resource.content))
     end
+    it 'includes an iframe resource' do
```
Comment (fotinakis) [inline spec/lib/percy/capybara/loaders/base_loader_spec.rb]: Awesome!

Can you add a new `context` at the top-level (same level as describe) for making sure this works in non-js enabled tests? Like this:

```ruby
context 'rack-test without javascript', type: :feature do
  # ... duplicate the `it` test here
end
```

File: lib/percy/capybara/loaders/sprockets_loader.rb [hunk 3]
```diff
@@ -25,9 +25,10 @@ def initialize(options = {})
         end
 
         def snapshot_resources
-          # When loading via Sprockets, all other resources are associated to the build, so the only
-          # snapshot resource to upload is the root HTML.
-          [root_html_resource]
+          resources = []
```
Comment (fotinakis) [inline lib/percy/capybara/loaders/sprockets_loader.rb]: Alternatively, just one line: `[root_html_resource] + get_iframes_resources`

File: lib/percy/capybara/loaders/base_loader.rb [hunk 4]
```diff
@@ -70,6 +70,34 @@ def current_path
 
           current_url
         end
+
+        # NOTES:
+        # - Doesn't handle multiple iframes with the same URL (`src` attribute)
+        # @private
+        def get_iframes_resources
```
Comment (fotinakis) [inline lib/percy/capybara/loaders/base_loader.rb]: Can you rename this to `iframe_resources` for consistency?

**Comment by fotinakis** [comment]
Back to you @christophemaximin, this is fantastic.

File: spec/lib/percy/capybara/loaders/base_loader_spec.rb (hunk 2 above)
Comment (christophemaximin) [inline spec/lib/percy/capybara/loaders/base_loader_spec.rb]: I do need to put that test on a proper level/context,
however if I write it without `js: true`, for consistency I'd have to also do this for https://github.com/percy/percy-capybara/blob/master/spec/lib/percy/capybara/loaders/base_loader_spec.rb#L4,
and write an sample Rack app to answer requests like https://github.com/percy/percy-capybara/blob/master/spec/lib/percy/capybara/loaders/sprockets_loader_spec.rb#L3
Should I actually do this or should I just keep the `js: true` like the `#root_html_resource` test?

Comment (fotinakis) [inline spec/lib/percy/capybara/loaders/base_loader_spec.rb]: Hmm, yes that is a bit of setup. My only concern here is that we haven't verified that the iframe logic is safe to run against RackTest (ie. when `js: false`, the default)—if iframe logic was behind a flag I'd be less concerned, but since we're enabling it by default it's more important to make sure it works in all environments.

Would you mind taking a stab at adding the test for this? Don't worry about adding a `#root_html_resource` test, just the iframe one is sufficient.

Quick stab:
```ruby
IFRAME_PATH = File.expand_path('../../client/testdata/test-iframe.html', __FILE__)

class RackAppWithIframe
  def self.call(env)
    [200, {}, File.read(IFRAME_PATH)]
  end
end

#...

  context 'Rack::Test', type: :feature do
    before(:each) { Capybara.app = RackAppWithIframe }
    after(:each) { Capybara.app = nil }
    # ... duplicate the `it` test here
  end
```

However! Note that `Rack::Test` is super weird and does not run a browser underneath, so there may not be any iframe source at all. I just want to make sure it runs without errors, not that it works correctly (I'll be surprised if it does). We will probably have to say that `js: true` is required for iframe support.

Comment (christophemaximin) [inline spec/lib/percy/capybara/loaders/base_loader_spec.rb]: Looks like we'll need `js: true` indeed, I get this with the Rack app:
`Capybara::NotSupportedByDriverError: Capybara::Driver::Base#within_frame`

Comment (fotinakis) [inline spec/lib/percy/capybara/loaders/base_loader_spec.rb]: Excellent, I'm glad we're adding this test otherwise all RackTests would have broke!

Let's just catch that error and silence it. Yeah?

Comment (fotinakis) [inline spec/lib/percy/capybara/loaders/base_loader_spec.rb]: To be clear, I'm talking about a _new_ test — this one is fine as is, just want to add a different one to make sure RackTests still run if there is an iframe in the HTML.

Comment (christophemaximin) [inline spec/lib/percy/capybara/loaders/base_loader_spec.rb]: Ah, got it, on it.

Comment (christophemaximin) [inline spec/lib/percy/capybara/loaders/base_loader_spec.rb]: What do you think about this: https://gist.github.com/christophemaximin/1016c25f437203c5902deded6635f771#file-diff-diff ?

Comment (fotinakis) [inline spec/lib/percy/capybara/loaders/base_loader_spec.rb]: LGTM! You could just expect `eq([])` instead of `.to_not raise_error`

**Comment by fotinakis** [comment]
Thanks so much for contributing this.

Released as `2.3.0.pre.beta` of the gem. Having a few customers test it out before tagging as 2.3.0.

**Comment by christophemaximin** [comment]
Thank you for your help and responsiveness!
Love Percy, been looking for an easy-to-integrate service like that for years, as I've always believed in the crucial importance of integrated visual testing; so I'm happy to do what I can to help :)
